lvmd/command/lvm.go

```diff
@@ -415,14 +423,86 @@ func (t *ThinPool) ListVolumes() ([]*LogicalVolume, error) {
 	return ret, nil
 }
 
+// FindVolume finds a named logical volume in this thin pool
+func (t *ThinPool) FindVolume(name string) (*LogicalVolume, error) {
```

Comment:
This function calls ListVolumes which loops over all the volumes in the vg and returns those in the thinpool. This function then loops over those volumes to find the matching name.
It would be more efficient to write a function which will list the volumes in the vg and check pool pool and name in the same loop.

```
func (t *ThinPool) FindVolume(name string) (*LogicalVolume, error) { {
	volumes, err := t.vg.ListVolumes()
	if err != nil {
		return nil, err
	}

	for _, volume := range volumes {
		if volume.name == name && volume.pool != nil && *volume.pool == t.name {
			return volume, nil
		}
	}
	return nil, ErrNotFound
}
```

Reply:
- ack, makes sense, since the lv names are unique the condition will be evaluated only once and reduces looping as well compared to existing implementation
- again, pasting the diff as not all code needs to be re-reviwed
``` diff
# gd thin-lv-3..temp
diff --git a/lvmd/command/lvm.go b/lvmd/command/lvm.go
index 0d030191..24aff2b9 100644
--- a/lvmd/command/lvm.go
+++ b/lvmd/command/lvm.go
@@ -425,12 +425,13 @@ func (t *ThinPool) ListVolumes() ([]*LogicalVolume, error) {

 // FindVolume finds a named logical volume in this thin pool
 func (t *ThinPool) FindVolume(name string) (*LogicalVolume, error) {
-       volumes, err := t.ListVolumes()
+       volumes, err := t.vg.ListVolumes()
        if err != nil {
                return nil, err
        }
+
        for _, volume := range volumes {
-               if volume.Name() == name {
+               if volume.name == name && volume.pool != nil && *volume.pool == t.name {
                        return volume, nil
                }
        }
@@ -585,11 +586,6 @@ func (l *LogicalVolume) Pool() (*ThinPool, error) {
        return l.vg.FindPool(*l.pool)
 }

-// GetPoolName returns a pointer to name of the logical volumes
-func (l *LogicalVolume) GetPoolName() *string {
-       return l.pool
-}
-
 // MajorNumber returns the device major number.
 func (l *LogicalVolume) MajorNumber() uint32 {
        return l.devMajor
diff --git a/lvmd/vgservice.go b/lvmd/vgservice.go
index da2ca18c..ded3aa49 100644
--- a/lvmd/vgservice.go
+++ b/lvmd/vgservice.go
@@ -71,7 +71,7 @@ func (s *vgService) GetLVList(_ context.Context, req *proto.GetLVListRequest) (*

        vols := make([]*proto.LogicalVolume, 0, len(lvs))
        for _, lv := range lvs {
-               if dc.Type == TypeThick && lv.GetPoolName() != nil {
+               if dc.Type == TypeThick && lv.IsThin() {
                        // do not send thin lvs if request is on TypeThick
                        continue
                }
```
- thanks @Yuggupta27 for pointing out existing `IsThin()` method.